source/request-response/subscription_manager.c

```diff
@@ -0,0 +1,624 @@
+        "request-response subscription manager - received a protocol adapter subscription event for ('" PRInSTR
+        "'), type %s, error_code %d(%s)",
+        AWS_BYTE_CURSOR_PRI(event->topic_filter),
+        aws_protocol_adapter_subscription_event_type_to_c_str(event->event_type),
+        event->error_code,
+        aws_error_debug_str(event->error_code));
+
+    if (event->event_type == AWS_PASET_SUBSCRIBE) {
+        AWS_FATAL_ASSERT(record->pending_action == ARRSPAT_SUBSCRIBING);
+
+        if (event->error_code == AWS_ERROR_SUCCESS) {
+            record->status = ARRSST_SUBSCRIBED;
+            s_emit_subscription_event(manager, record, ARRSET_SUBSCRIPTION_SUBSCRIBE_SUCCESS);
+        } else {
+            s_emit_subscription_event(manager, record, ARRSET_SUBSCRIPTION_SUBSCRIBE_FAILURE);
+        }
+    } else if (event->event_type == AWS_PASET_UNSUBSCRIBE) {
+        AWS_FATAL_ASSERT(record->pending_action == ARRSPAT_UNSUBSCRIBING);
+
+        if (event->error_code == AWS_ERROR_SUCCESS) {
+            record->status = ARRSST_NOT_SUBSCRIBED;
+            s_emit_subscription_event(manager, record, ARRSET_SUBSCRIPTION_ENDED);
+        }
+    }
+
+    record->pending_action = ARRSPAT_NOTHING;
+
+    s_aws_rr_subscription_record_log_invariant_violations(record);
+}
+
+static int s_rr_activate_idle_subscriptions_wrapper(void *context, struct aws_hash_element *elem) {
+    struct aws_rr_subscription_record *record = elem->value;
+    struct aws_rr_subscription_manager *manager = context;
+
+    s_rr_activate_idle_subscription(manager, record);
+
+    s_aws_rr_subscription_record_log_invariant_violations(record);
+
+    return AWS_COMMON_HASH_TABLE_ITER_CONTINUE;
+}
+
+static void s_activate_idle_subscriptions(struct aws_rr_subscription_manager *manager) {
+    aws_hash_table_foreach(&manager->subscriptions, s_rr_activate_idle_subscriptions_wrapper, manager);
+}
+
+static int s_apply_session_lost_wrapper(void *context, struct aws_hash_element *elem) {
+    struct aws_rr_subscription_record *record = elem->value;
+    struct aws_rr_subscription_manager *manager = context;
+
+    if (record->status == ARRSST_SUBSCRIBED) {
+        record->status = ARRSST_NOT_SUBSCRIBED;
+        s_emit_subscription_event(manager, record, ARRSET_SUBSCRIPTION_ENDED);
+
+        if (record->pending_action != ARRSPAT_UNSUBSCRIBING) {
+            s_aws_rr_subscription_record_destroy(record);
+            return AWS_COMMON_HASH_TABLE_ITER_CONTINUE | AWS_COMMON_HASH_TABLE_ITER_DELETE;
+        }
+    }
+
+    return AWS_COMMON_HASH_TABLE_ITER_CONTINUE;
+}
+
+static void s_apply_session_lost(struct aws_rr_subscription_manager *manager) {
+    aws_hash_table_foreach(&manager->subscriptions, s_apply_session_lost_wrapper, manager);
+}
+
+void aws_rr_subscription_manager_on_protocol_adapter_connection_event(
+    struct aws_rr_subscription_manager *manager,
+    const struct aws_protocol_adapter_connection_event *event) {
+
+    AWS_LOGF_DEBUG(
```

Comment:
Trivial: We should probably move this log into the if statements below `if (event->event_type == AWS_PACET_CONNECTED)` and exclude the "joined_session %d" for disconnected connection events for slightly cleaner logging. Or at least explicitly set the `event->joined_session` to false when creating the aws_protocol_adapter_connection_event struct before it is passed here rather than potentially taking on whatever that value was on a non-zeroed struct.